Allow arbitrary UI code in the choices for radio buttons and checkbox group input

bborgesr opened the pull request:
Example app:

```r
ui <- fluidPage(
  sidebarLayout(
    sidebarPanel( 
      radioButtons("normal", "Normal radio buttons:",
        c("Normal" = "norm", "Uniform" = "unif",
         "Log-normal" = "lnorm", "Exponential" = "exp")),
      hr(),
      radioButtons("fancy", "Fancy radio buttons:",
      choiceNames = list(tags$span(HTML("Normal &mu;<sub>1</sub>")),
        icon("calendar", "fa-3x"),
        icon("bed"),
        icon("cog", lib = "glyphicon"),
        img(src = "https://www.rstudio.com/wp-content/uploads/2016/09/RStudio-Logo-Blue-Gray-250.png")),
      choiceValues = c("html", "icon1", "icon2", "icon3", "img"))),
    mainPanel(textOutput("rb1"), textOutput("rb2"))
  )
)

server <- function(input, output, session) {
  output$rb1 <- renderText({
    paste("You chose", input$normal)
  })
  
  output$rb2 <- renderText({
    paste("You chose", input$fancy)
  })
}

shinyApp(ui, server)
```

bborgesr commented:
Fixes #1348 and fixes #1437

jcheng5 commented:
I'd prefer `choiceNames` and `choiceValues` instead of `choicesNames` and `choicesValues`.

I don't like that `choices`, `choicesNames`, and `choicesValues` propagate beyond `checkChoicesArgs`; I think `choices` should go away, since it can be thought of as a shorthand for `choicesNames <- names(choices)` and `choicesValues <- unname(choices)`. Then the rest of the code (like `validateSelected`) can just deal with one way of specifying these values, instead of two.

I tend to think of functions that are named `checkFoo` and `validateFoo` as simply returning `TRUE` or `FALSE`, or throwing warnings or errors. In these cases we're using them to do normalization. Not a big deal and you don't have to change the names in this PR but maybe something to keep in mind moving forward.

(Don't forget the inspecting/merging of the API branch needs to happen before the conference, this PR can be merged after the conference...)

wch commented on R/input-utils.R:
```diff
@@ -2,30 +2,65 @@ controlLabel <- function(controlName, label) {
+    if (lenNames != 0 || lenVals != 0) {
+      warning("Using `choices` argument; ignoring `choicesNames` and
+                `choicesValues`.")
+      choicesNames = NULL
```
Should be `<-`.

bborgesr commented on R/input-utils.R:
```diff
@@ -2,30 +2,65 @@ controlLabel <- function(controlName, label) {
+      stop("`choicesNames` and `choicesValues` must not be named.")
+    }
+  } else {
+    if (lenNames != 0 || lenVals != 0) {
```
check for NULL instead

bborgesr commented on R/input-utils.R:
```diff
@@ -2,30 +2,65 @@ controlLabel <- function(controlName, label) {
+    if (lenNames != lenVals) {
+      stop("`choicesNames` and `choicesValues` must have the same length.")
+    }
+    if (!is.null(names(choicesNames)) || !is.null(names(choicesValues))) {
```
use `anyNamed` instead

bborgesr commented:
I think I've addressed all points and this should be good to squash-merge. 

The one possible hiccup is what we're breaking backward compatibility (but only for unexported functions!) as we change the name of some of the internal utility functions (switch from "validate"/"check" prefixes to "normalize", per Joe's comment above) and their expected arguments (pass in `choiceNames` and `choiceValues` instead of `choices`).

jcheng5 commented:
Maybe we can review this together tomorrow. It's been so long I don't remember :)

wch commented on R/input-utils.R:
```diff
@@ -2,30 +2,62 @@ controlLabel <- function(controlName, label) {
 # refers to values. Below is a function for backward compatibility. It also
 # coerces the value to `character`.
-validateSelected <- function(selected, choices, inputId) {
+normalizeSelected <- function(selected, inputId, choiceNames, choiceValues) {
```
Since 0.10.0 (which was released in June 2014) users have gotten a warning if `selected` was a name instead of a value. So I think it's safe to assume that no one is doing that anymore. And that means that we can remove this function entirely. We will need to do `selected <- as.character(selected)` where we previously called `validateSelected`, for the reasons in the comment directly below.

wch commented on R/update-input.R:
```diff
@@ -452,16 +452,18 @@ updateSliderInput <- function(session, inputId, label = NULL, value = NULL,
-                               type = 'checkbox') {
-  if (!is.null(choices))
-    choices <- choicesWithNames(choices)
+                               selected = NULL, inline = FALSE, type = NULL,
```
It would be good to error if `type` is NULL.

wch commented on R/update-input.R:
```diff
@@ -452,16 +452,18 @@ updateSliderInput <- function(session, inputId, label = NULL, value = NULL,
+      args$choiceNames, args$choiceValues)
 
-  options <- if (!is.null(choices)) {
+  options <- if (!is.null(args$choiceValues)) {
```
This should be `if (!is.null(choices) || !is.null(choiceNames) || !is.null(choiceValues))`.

wch commented on R/input-utils.R:
```diff
@@ -2,30 +2,62 @@ controlLabel <- function(controlName, label) {
+    choiceValues <- unname(choices)
+  }
+
+  return(list(choiceNames = choiceNames, choiceValues = choiceValues))
```
If `choices`, `choicesNames`, and `choicesValues` are all NULL, this should return a list like:

```
list(choiceNames = NULL, choiceValues = NULL)
```

And also have a test for it.

wch replied:
Also, `choiceNames` and `choiceValues` should always be lists (or NULL), and not character vectors.

wch commented on NEWS.md:
```diff
@@ -5,6 +5,8 @@ shiny 1.0.0.9001
 
 ### Breaking changes
 
+* The functions `radioButtons()`, `checkboxGroupInput()` and `selectInput()` (and the corresponding `updateXXX()` functions) no longer accept a `selected` argument whose value is the name of a choice, instead of the value of the choice. This feature had been soft-deprecated since Shiny 0.10 (printing a warning message, but still trying to match the name to the right choice) and it's now been comp ...
```
It would be more accurate to say that it had been deprecated (not soft-deprecated) since 0.10.0.
